api/agent/agent.go

```diff
@@ -737,114 +745,183 @@ func newSizerRespWriter(max uint64, rw http.ResponseWriter) http.ResponseWriter
 
 func (s *sizerRespWriter) Write(b []byte) (int, error) { return s.w.Write(b) }
 
-func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state ContainerState) {
+func (a *agent) waitInitialize(ctx context.Context, caller slotCaller, evictor *EvictToken, udsWait <-chan error) (bool, error) {
+	for {
+		select {
+		case err := <-udsWait:
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		case <-ctx.Done(): // container shutdown
+			return false, nil
+		case <-a.shutWg.Closer(): // agent shutdown
+			return false, nil
+		case <-caller.done: // original caller disconnected or serviced by another container?
+			evictor.SetEvictable(true)
+			caller.done = nil // block 'doneChan' after this point
+		case <-evictor.C: // eviction
+			return false, nil
+		}
+	}
+}
+
+// Try to queue an error to the error channel if possible.
+func tryQueueErr(err error, ch chan error) error {
+	if err != nil {
+		select {
+		case ch <- err:
+		default:
+		}
+	}
+	return err
+}
+
+func (a *agent) runHot(ctx context.Context, caller slotCaller, call *call, tok ResourceToken, state ContainerState) {
 	// IMPORTANT: get a context that has a child span / logger but NO timeout
 	// TODO this is a 'FollowsFrom'
 	ctx = common.BackgroundContext(ctx)
 	ctx, span := trace.StartSpan(ctx, "agent_run_hot")
 	defer span.End()
 
-	// IMPORTANT: evict token is deleted *after* resource token in defer statements below.
-	// This ordering allows resource token to be freed first, which means once evict token
-	// is deleted, eviction is considered to be completed.
+	var container *container
+	var cookie drivers.Cookie
+	var waiter drivers.WaitResult
+	var runRes drivers.RunResult
```

Comment:
i think `runRes` and `waiter` can be eliminated from this list of upfront declarations, the only use appears to be for a log line in a separate goroutine which can go in the same thread and to not use `:=` in the other (at least as someone reading this i have to go back and forth over usage to be able to reason about scope these are being used in to figure out what's going on, and it's a lot of cognitive overhead for something that maybe doesn't need to be). reducing these in general would be nice, if possible. tracking back and forth between this and goroutine isn't very fun. a lot of the defers could be on their own, at least `container.Close` and `cookie.Close`, after checking the err on the method that produces them. think that covers them all.

Reply:
i kinda agree that seeing the ordering inside of the defer is nice instead of having to read defers backwards, so remove 1/2 this comment I guess. the defer itself sticks out (ie read about what happens at the end before doing the things) but overall seems ok

Reply:
This got slightly better now I think.

Reply:
Our tear down ordering is tricky, so I really felt the need to spell it out there and avoid any bugs if we shuffle/reorder things in runHot() body.